### PR TITLE
Emily

### LINKED ISSUE #620: [Concepts 4] useOfGraphs: Visuals, Context
#### _1.md_
* add visual
* add an IN DEPTH example (ex: airports, facebook/social media)
* why would you want to traverse this example graph?
* what might a use case be?

### LINKED ISSUE #623: [Concept 4] realLifeApplication: Context
#### _1.md_
* go THROUGH an entire example
* run dijkstra's on the example
* time and space complexity reminder

### LINKED ISSUE #924: [Concepts 4] Search For an Element in a Linked List: Context
#### _activity_
* write in the iterative solution. but do mention that it is not commonly used so that "###Search an element in a Linked List (Iterative and Recursive)" is not misleading

### LINKED ISSUE #926: [Activity 2.2.1] DFS: Context
#### _activity_
* please remove all references of dead-end. it can be used figuratively at most ONCE
* talk about auxiliary DS
* talk about time and space complexity

### LINKED ISSUE #930: [Concept 4] Undirected Graphs: Styling, Context
#### _activity_
* Can the edges be weighted?
* typos

### LINKED ISSUE #931: [Concept 4] Dictionary vs List: Context
#### _activity_
* add big O comparison (space vs time in this case)

### LINKED ISSUE #934: [Concept 4] Dijkstra's Undirected: Context
#### _activity_
* how is this different from using it on a directed graph?

### PR DESCRIPTION
**Pull request by emily-gui**
Closes #620 
Closes #623 
Closes #924 
Closes #926 
Closes #930 
Closes #931 
Closes #934 
...

Changes proposed in this pull request:
* added concepts from airtable
* edited cards to requested changes
* added visual and explanations
* fixed typos

@mxthu313

### CODE REVIEW

File: Module4_Concepts/Depth-FirstSearch(DFS).md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+<!--title={Depth First Search (DFS)}-->
+
+# Depth-First Search 
+
+Depth-First Search(DFS) is one of our data searching algorithms we use in Python to navigate Data structures, Graphs, and Search Trees. DFS is best explained with a visual aide. 
+
+![]( https://upload.wikimedia.org/wikipedia/commons/thumb/1/1f/Depth-first-tree.svg/450px-Depth-first-tree.svg.png)
+
+DFS starts at a root point and branches on from that point. DFS continues down one path as far as possible before backtracking to the root point and taking an alternate path.
```
Comment (mxthu313) [inline Module4_Concepts/Depth-FirstSearch(DFS).md]: use node/vertex instead of point

File: Module4_Concepts/Depth-FirstSearch(DFS).md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+# Depth-First Search 
+
+Depth-First Search(DFS) is one of our data searching algorithms we use in Python to navigate Data structures, Graphs, and Search Trees. DFS is best explained with a visual aide. 
+
+![]( https://upload.wikimedia.org/wikipedia/commons/thumb/1/1f/Depth-first-tree.svg/450px-Depth-first-tree.svg.png)
+
+DFS starts at a root point and branches on from that point. DFS continues down one path as far as possible before backtracking to the root point and taking an alternate path. 
+
+If we trace the numbers in the visual aide, from 1-12, we get a good sense of idea of how DFS can navigates data and graphs.
```
Comment (mxthu313) [inline Module4_Concepts/Depth-FirstSearch(DFS).md]: grammar

File: Module4_Concepts/Depth-FirstSearch(DFS).md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+   Circle 5 -> Circle 3 -> Circle 2
+
+5. Circle 2 has the available path of Circle 6.
+
+6. We continue this pattern until we have searched the entire data structure, tree, or graph. 
+
+To explain it in terms of python, we follow these rules.
+
+- **Rule 1** − Visit the adjacent unvisited vertex. Mark it as visited. Display it. Push it in a stack.
```
Comment (mxthu313) [inline Module4_Concepts/Depth-FirstSearch(DFS).md]: why a stack? what is the significance of the vertices that are stored in the stack?

File: Module4_Concepts/Depth-FirstSearch(DFS).md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+print("Following is DFS from (starting from vertex 2)") 
+g.DFS(2) 
+
+# This code is contributed by Neelam Yadav 
+```
+
+---
+
+The time complexity of DFS is O(N^M), if N is the number of nodes in a tree and M is the maximum depth of the tree.
```
Comment (mxthu313) [inline Module4_Concepts/Depth-FirstSearch(DFS).md]: hmmmmm. I don't think the complexities are right.

File: Module4_Concepts/Dictionary vs List.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+```python
+Dict['a'] # which is equivalent to accessing the value 2
+```
+
+This kind of indexing is not possible in lists. It's useful for quick retrieval or checking whether a key exists or not, or what's the corresponding value. The same operation in a list would involve a linear search of the complete list in worst case thus making a list inefficient for this kind of operations.
+
+---
+
+This is a table comparing the time and space complexities of lists and dictionaries for different functions:
```
Comment (mxthu313) [inline Module4_Concepts/Dictionary vs List.md]: can you make a custom graphic instead? I don't trust this to be perfectly rendered. You can literally take a screenshot of a google doc. or a screenshot of how it is displayed in your screen :)

File: Module4_Concepts/RealLifeApplication.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+<!--title={Dijkstra's Algorithm Real Life Application}-->
+
+Suppose that our GPS (Global Positioning System) uses Dijkstra's algorithm to find the shortest route between one location to another, and to demonstrate how this GPS works, let's imagine that graph below is a geographical map—the nodes representing the cities on the map, the edges representing the roads between cities, and the weights of the edges representing distance:
+
+![Dijkstra Algorithm in Java | Baeldung](https://www.baeldung.com/wp-content/uploads/2017/01/initial-graph.png)
```
Comment (mxthu313) [inline Module4_Concepts/RealLifeApplication.md]: i like this walk through :)

Comment (emily-gui) [reply Module4_Concepts/RealLifeApplication.md]: yayy thanks! :>

File: Module4_Concepts/Dictionary vs List.md (hunk 5 above)
Comment (mxthu313) [reply Module4_Concepts/Dictionary vs List.md]: can you please place that in an "image" folder in the concept folder? And use the  tag

**Comment by mxthu313** [comment]
beautiful :)

**Pull request merged**